Fix our go.mod version string parser regex pattern

eskultety opened the pull request:
We currently don't allow prerelease version identifiers (e.g. beta/rc, etc.) and commentaries on the version string line (e.g. `go 1.21.0//comment`.

This PR mainly fixes a nasty bug where we fell back to picking an older Go version because of ^this which resulted in a very strange error trying to build buildah from git:

```
2024-05-29 15:48:50,010 INFO Fetching the gomod dependencies at subpath buildah
2024-05-29 15:48:50,011 INFO Fetching the gomod dependencies at the "buildah" directory
2024-05-29 15:48:52,913 INFO go.mod reported versions: '1.20'[go], '-'[toolchain] (cachi2 enforced)
2024-05-29 15:48:52,927 INFO Using Go release: go1.20
2024-05-29 15:48:52,928 INFO Vendoring the gomod dependencies
2024-05-29 15:49:17,954 INFO Retrieving the list of packages
2024-05-29 15:49:23,014 ERROR RuntimeError: Package parent module was not found
Traceback (most recent call last):
```

even though buildah uses 1.21 already. This PR also updates some unit tests and adds debug logs to keep better track of dealing with version string parsing.

# Maintainers will complete the following section

- [ ] Commit messages are descriptive enough
- [ ] Code coverage from testing does not decrease and new code is covered
- [ ] Docs updated (if applicable)
- [ ] Docs links in the code are still valid (if docs were updated)

**Note:** if the contribution is external (not from an organization member), the CI
pipeline will not run automatically. After verifying that the CI is safe to run:

- [approve GitHub Actions workflows][approve-gh-actions] by clicking a button
- approve the Red Hat Trusted App Pipeline container build by commenting `/ok-to-test`
  (as is the [standard for Pipelines as Code][pac-running-pipeline])

[approve-gh-actions]: https://docs.github.com/en/actions/managing-workflow-runs/approving-workflow-runs-from-public-forks
[pac-running-pipeline]: https://pipelinesascode.com/docs/guide/running/#running-the-pipelinerun

eskultety commented:
**Note that commits 5 and 6 are going to be squashed together before merging, I split them only for review purposes to make the regex changes easier to track.**

eskultety commented:
Since v1:
- added a commit that further tweaks how and when we report what version we parsed out of  the `go.mod` file. I verified that this PR fixes the problem reported in the description based on the original buildah submodule ref used here: https://github.com/konflux-ci/buildah-container/commit/588f5baf3d2e4d8e4ffa02d7ed604c6543874dff

taylormadore reviewed (approved):
LGTM :+1: 

Thanks for improving the logging around this

eskultety commented on cachi2/core/package_managers/gomod.py:
```diff
@@ -677,20 +678,29 @@ def _get_gomod_version(go_mod_file: RootedPath) -> Tuple[Optional[str], Optional
+        for i, line in enumerate(f):
             if not go_version and (match := re.match(go_pattern, line)):
                 go_version = match.group("ver")
+                log.debug(f"Matched Go version {go_version} on go.mod line {i}: '{line}'")
```
In this particular case I wouldn't call it lazy as there's inherently nothing lazy about it, we're constructing the string almost at the same time (right before the method call vs within the method itself).
What performance optimization are you talking  about exactly? f-strings are known to be the fastest string formatting method in Python (not even needed to compare with Templates):

```
import timeit

expr = """
amet = 'amet'
elit = 'elit'
N = 42

s = 'Lorem ipsum dolor sit %s, consectetur adipiscing %s. Vestibulum non %d eget ipsum fermentum faucibus' % (amet, elit, N)
"""
print(timeit.timeit(expr, number=10000))


expr = """
amet = 'amet'
elit = 'elit'
N = 42

s = f'Lorem ipsum dolor sit {amet}, consectetur adipiscing {elit}. Vestibulum non {N} eget ipsum fermentum faucibus'
"""

print(timeit.timeit(expr, number=10000))
-----------------------------------------------------
$ python3 script.py
0.002215377000538865
0.001156313000137743
```

I only noticed the old style string formatting after you pointed it out and we're heavily inconsistent, that's true. That said

```
$ git grep " f\".*\"" | wc -l
492
$ git grep "\".**\%s.*\"" | wc -l
89
```
turns the scales in favour of f-strings. Even the official docs acknowledges support for other formatting options and states that they focus on the old style formatting due to backwards compatibility

> As you can see, merging of variable data into the event description message uses the old, %-style of string formatting. **This is for backwards compatibility:** the logging package pre-dates newer formatting options such as [str.format()](https://docs.python.org/3/library/stdtypes.html#str.format) and [string.Template](https://docs.python.org/3/library/string.html#string.Template).

So my take on this is that we should always use f-strings everywhere and fix the rest of the old style string formatting occurrences to instate consistency.

MartinBasti replied:
No formatting is faster than any formatting, your proof is not valid for logging use case.

If DEBUG level is not enabled, logging module won't format anything, it just skip, no string is created.

```
log.debug("this %s string won't be even created", "example")  # note, it's not `str % "val"`

log.debug(f"this {val} string will be crated and just thrown away if debug level is not enabled")
```

MartinBasti replied:
https://github.com/python/cpython/blob/22c9d9c1fcc3bb6186524330b169eda6df450f1b/Lib/logging/__init__.py#L1506

eskultety replied:
That's true, are we going to apply this reasoning to all strings we have in the code base and dispute consistency or are we okay with the miniscule time it really takes to format the string?

MartinBasti replied:
it's consistent, all log entries should use that. It's documented as best practice for python, even pylint has check for it.

Other strings can be f-strings

eskultety replied:
Fair enough.  For the record, @MartinBasti shared the following article with me privately which mentioned that f-strings in logging can pose an unnecessary performance penalty when formatting large objects which is something I didn't consider.
I updated the log formatting strings in this PR. I'll post a follow up on the rest of logging strings to reinstate consistency.

eskultety commented:
> **Note that commits 5 and 6 are going to be squashed together before merging, I split them only for review purposes to make the regex changes easier to track.**

As advertised, I squashed the two before merging.
